libraries/AP_Proximity/AP_Proximity_LD06.cpp

```diff
@@ -0,0 +1,185 @@
+    } else {
+        set_status(AP_Proximity::Status::Good);
+    }
+}
+
+// Returns the maximum distance (m) that the sensor supports
+float AP_Proximity_LD06::distance_max() const
+{
+    return 12.0f;  // LD06 has a maximum range of 12m
+}
+
+// Returns the minimum distance (m) that the sensor supports
+float AP_Proximity_LD06::distance_min() const
+{
+    return 0.02f;  // LD06 has a minimum range of 2 cm (0.02 m)
+}
+
+// Called repeatedly to get the readings at the current instant
+void AP_Proximity_LD06::get_readings()
+{
+    if (_uart == nullptr) {
+        return;
+    }
+
+    // Store the number of bytes available on the UART input
+    uint32_t nbytes = _uart->available();
+
+    // Loops through all bytes that were received
+    while (nbytes-- > 0) {
+
+        // Gets and logs the current byte being read
+        const uint8_t c = _uart->read();
+
+        // Stores the byte in an array if the byte is a start byte or we have already read a start byte
+        if (c == LD_START_CHAR || _response_data) {
+
+            // Sets to true if a start byte has been read, default false otherwise
+            _response_data = true;
+
+            // Stores the next byte in an array
+            _response[_byte_count] = c;
+            _byte_count++;
+
+            // The response should have 47 bytes in total
+            if (_byte_count == sizeof(_response)) {
+
+                set_status(AP_Proximity::Status::Good);
+
+                uint32_t current_ms = AP_HAL::millis();
+
+                // Stores the last distance taken, used to reduce number of readings taken
+                if (_last_distance_received_ms == 0 || _last_distance_received_ms > current_ms) {
+                    _last_distance_received_ms = current_ms;
+                }
+
+                // Checks if it has been more than 50 ms since the last reading - results in 20 Hz sampling rate
+                if (_last_distance_received_ms + 50 > current_ms) {
+                    _last_distance_received_ms =  current_ms;
+                    
+                    // Updates the temporary boundary and passes off the completed data
+                    parse_response_data();
+                    _temp_boundary.update_3D_boundary(boundary);
+                    _temp_boundary.reset();
+                }
+
+                // Resets the bytes read and whether or not we are reading data to accept a new payload
+                _byte_count = 0;
+                _response_data = false;
+            }
+        }
+    }
+}
+
+// Parses the data packet received from the LiDAR
+void AP_Proximity_LD06::parse_response_data()
+{ 
+
+    // Data interpretation based on:
+    // http://wiki.inno-maker.com/display/HOMEPAGE/LD06?preview=/6949506/6949511/LDROBOT_LD06_Development%20manual_v1.0_en.pdf
+
+    // Second byte in array stores length of data - not used but stored for debugging
+    // const uint8_t data_length = _response[START_DATA_LENGTH];
```

Comment:
We should likely be checking the length field rather than assuming here and in the part above. As one day there may be more messages types?